### PR TITLE
[samplefocus] Add new extractor

### LINKED ISSUE #27741: SampleFocus support
## Checklist

- [x] I'm reporting a new site support request
- [x] I've verified that I'm running youtube-dl version **2021.01.08**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

- Single sample: https://samplefocus.com/samples/dababy-style-bass-808
- Search query: https://samplefocus.com/tag/bass

## Description

If you have ever used DevTools, you'll be able to get the MP3s within 3 minutes. (play the sample and look at the Network tab) Getting WAVs is impossible without finding a vulnerability.

### PR DESCRIPTION
**Pull request by Isaac-the-Man**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added new extractor for the site [https://samplefocus.com/](https://samplefocus.com/). Closes #27741.

### CODE REVIEW

File: youtube_dl/extractor/samplefocus.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title')
+        tb = self._og_search_thumbnail(webpage) or extract_attributes(get_element_by_class('waveform')).get('src')
+
+        mp3_url = self._html_search_regex(
+            r'<meta itemprop="contentUrl" content="?(.+?)"?>',
+            webpage, 'mp3 url') or extract_attributes(get_element_by_id('sample_mp3')).get('value')
```
Comment (adrianheine) [inline youtube_dl/extractor/samplefocus.py:41]: `_html_search_regex` will just error out if you don't provide a `default` or set `fatal=False`.

File: youtube_dl/extractor/samplefocus.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        return {
+            'id': video_id,
+            'title': title,
+            'url': mp3_url,
+            'ext': 'mp3',
+            'thumbnail': tb,
+            'description': self._html_search_meta('description', webpage),
+            'license': self._html_search_regex(r'<a href="/license">(.+?)</a>', webpage, 'license')
```
Comment (adrianheine) [inline youtube_dl/extractor/samplefocus.py:50]: I suggest `fatal=False`

File: youtube_dl/extractor/samplefocus.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }, {
+        'url': 'https://samplefocus.com/samples/young-chop-kick',
+        'only_matching': True
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title')
```
Comment (adrianheine) [inline youtube_dl/extractor/samplefocus.py:36]: I suggest `default=video_id` in the `_html_search_regex` call.

**Review by adrianheine — CHANGES_REQUESTED** [review]

**Review by adrianheine — APPROVED** [review]

File: youtube_dl/extractor/samplefocus.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        'url': 'https://samplefocus.com/samples/young-chop-kick',
+        'only_matching': True
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title', default=video_id)
+        tb = self._og_search_thumbnail(webpage) or extract_attributes(get_element_by_class('waveform')).get('src')
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:37]: - extract mandatory information first.
- incorrect fallback.

File: youtube_dl/extractor/samplefocus.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+        mp3_url = self._html_search_regex(
+            r'<meta itemprop="contentUrl" content="?(.+?)"?>',
+            webpage, 'mp3 url', fatal=False) or extract_attributes(get_element_by_id('sample_mp3')).get('value')
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:41]: - incorrect fallback.
- use `_hidden_inputs` method.

File: youtube_dl/extractor/samplefocus.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title', default=video_id)
+        tb = self._og_search_thumbnail(webpage) or extract_attributes(get_element_by_class('waveform')).get('src')
+
+        mp3_url = self._html_search_regex(
+            r'<meta itemprop="contentUrl" content="?(.+?)"?>',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:40]: - use `sample_mp3` as the primary source.
- match only up to mp3 extension.

File: youtube_dl/extractor/samplefocus.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+from .common import InfoExtractor
+from ..utils import (
+    get_element_by_class,
+    extract_attributes,
+    get_element_by_id)
+
+
+class SampleFocusIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?samplefocus\.com/samples/(?P<id>[\w,\-,0-9]+)'
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:12]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?samplefocus\.com/samples/(?P<id>[^/?&#]+)'
```

File: youtube_dl/extractor/samplefocus.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import (
+    get_element_by_class,
+    extract_attributes,
+    get_element_by_id)
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:8]: - Alphabetic.
- Vertical Hanging Indent.

File: youtube_dl/extractor/samplefocus.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    _VALID_URL = r'https?://(?:www\.)?samplefocus\.com/samples/(?P<id>[\w,\-,0-9]+)'
+    _TESTS = [{
+        'url': 'https://samplefocus.com/samples/lil-peep-sad-emo-guitar',
+        'md5': '48c8d62d60be467293912e0e619a5120',
+        'info_dict': {
+            'id': 'lil-peep-sad-emo-guitar',
+            'ext': 'mp3',
+            'title': 'Lil Peep Sad Emo Guitar',
+            'description': 'Listen to Lil Peep Sad Emo Guitar. Royalty-Free sound that is tagged as electric guitar, emo, guitar, and lil peep. Download for FREE + discover 1000\'s of sounds.',
+            'thumbnail': r're:^https?://.*\.png',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:21]:
```suggestion
            'thumbnail': r're:^https?://.+\.png',
```

File: youtube_dl/extractor/samplefocus.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+class SampleFocusIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?samplefocus\.com/samples/(?P<id>[\w,\-,0-9]+)'
+    _TESTS = [{
+        'url': 'https://samplefocus.com/samples/lil-peep-sad-emo-guitar',
+        'md5': '48c8d62d60be467293912e0e619a5120',
+        'info_dict': {
+            'id': 'lil-peep-sad-emo-guitar',
+            'ext': 'mp3',
+            'title': 'Lil Peep Sad Emo Guitar',
+            'description': 'Listen to Lil Peep Sad Emo Guitar. Royalty-Free sound that is tagged as electric guitar, emo, guitar, and lil peep. Download for FREE + discover 1000\'s of sounds.',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:20]: `md5:`

File: youtube_dl/extractor/samplefocus.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+        mp3_url = self._html_search_regex(
+            r'<meta itemprop="contentUrl" content="?(.+?)"?>',
+            webpage, 'mp3 url', fatal=False) or extract_attributes(get_element_by_id('sample_mp3')).get('value')
+
+        return {
+            'id': video_id,
+            'title': title,
+            'url': mp3_url,
+            'ext': 'mp3',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:47]: will be extracted from the URL.

File: youtube_dl/extractor/samplefocus.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            'thumbnail': tb,
+            'description': self._html_search_meta('description', webpage, fatal=False),
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:49]: `fatal=False` is already the default.

**Review by remitamine — CHANGES_REQUESTED** [review]
read [coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and apply them.

File: youtube_dl/extractor/samplefocus.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    }, {
+        'url': 'https://samplefocus.com/samples/young-chop-kick',
+        'only_matching': True
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title', default=video_id)
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:32]: the fallback will never be reached this way.

File: youtube_dl/extractor/samplefocus.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+        mp3_url = self._html_search_regex(
+            r'<input type="hidden" name="sample_mp3" id="sample_mp3" value="(.+\.mp3\?[0-9]+)" class="sample-mp3" />',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:35]: use `_hidden_inputs` method.

File: youtube_dl/extractor/samplefocus.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+            webpage, 'mp3', fatal=False) or self._html_search_regex(
+            r'<meta itemprop="contentUrl" content="?(.+\.mp3\?[0-9]+)"?>',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:37]: will easily match outside the element.

File: youtube_dl/extractor/samplefocus.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+            webpage, 'mp3 url')
+
+        tb = self._og_search_thumbnail(webpage) or self._html_search_regex(
+            r'<img class="waveform" style="background-color: #(?:[\w,0-9]{6})" src="(.+?)" />',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:41]: - does not match the correct image.
- you're not following the coding conventions.

**Review by remitamine — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/samplefocus.py (hunk 14 above)
Comment (Isaac-the-Man) [reply youtube_dl/extractor/samplefocus.py:35]: values returned from `_hidden_inputs` will be overwritten by the last value if multiple `<inputs type="hidden"/>` are present.

File: youtube_dl/extractor/samplefocus.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage, fatal=False) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title', default=video_id)
+
+        mp3_url = self._html_search_regex(
+            r'<input[^>]+type="hidden"[^>]+id=(?:["\'])sample_mp3(?:["\'])[^>]+value=(?:["\'])(.+\.mp3\?[0-9]+)(?:["\'])',
+            webpage, 'mp3', fatal=False) or self._html_search_regex(
+            r'<meta[^>]+itemprop=(?:["\'])contentUrl(?:["\'])[^>]+content=(?:["\'])?(.+\.mp3\?[0-9]+)(?:["\'])?',
+            webpage, 'mp3 url')
+        # print(mp3_url)
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:39]: remove.

File: youtube_dl/extractor/samplefocus.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage, fatal=False) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title', default=video_id)
+
+        mp3_url = self._html_search_regex(
+            r'<input[^>]+type="hidden"[^>]+id=(?:["\'])sample_mp3(?:["\'])[^>]+value=(?:["\'])(.+\.mp3\?[0-9]+)(?:["\'])',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:35]: - i don't think that `type="hidden"` is important.
- checking for ext is not needed here.

File: youtube_dl/extractor/samplefocus.py [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    }, {
+        'url': 'https://samplefocus.com/samples/young-chop-kick',
+        'only_matching': True
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage, fatal=False) or self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title', default=video_id)
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:32]: split long lines.

File: youtube_dl/extractor/samplefocus.py [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+        mp3_url = self._html_search_regex(
+            r'<input[^>]+type="hidden"[^>]+id=(?:["\'])sample_mp3(?:["\'])[^>]+value=(?:["\'])(.+\.mp3\?[0-9]+)(?:["\'])',
+            webpage, 'mp3', fatal=False) or self._html_search_regex(
+            r'<meta[^>]+itemprop=(?:["\'])contentUrl(?:["\'])[^>]+content=(?:["\'])?(.+\.mp3\?[0-9]+)(?:["\'])?',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:37]: - the part after `.mp3` is does have to be matched exactly(ex: if it changes to `?foo=bar` it should still be matched).
- the previous comment for this part still not addressed(matching outside element).

File: youtube_dl/extractor/samplefocus.py [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+            webpage, 'mp3', fatal=False)
+
+        return {
+            'id': video_id,
+            'title': title,
+            'url': mp3_url,
+            'ext': 'mp3',
+            'thumbnail': tb,
+            'description': self._html_search_meta('description', webpage),
+            'license': self._html_search_regex(r'<a href="/license">(.+?)</a>', webpage, 'license')
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:52]: the coding conventions should be applied to whole extractor.

**Review by remitamine — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/samplefocus.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage, fatal=False) or self._html_search_regex(
+            r'<h1>(.+?)</h1>', webpage, 'title', default=video_id)
+
+        mp3_url = self._html_search_regex(
+            r'<input[^>]+id=(?:["\'])sample_mp3(?:["\'])[^>]+value=(?:["\'])(.+\?[0-9]+)(?:["\'])[^>]*>',
+            webpage, 'mp3', fatal=False) or self._html_search_regex(
+            r'<meta[^>]+itemprop=(?:["\'])contentUrl(?:["\'])[^>]+content=(?:["\'])(?:["\'])[^>]*?>',
```
Comment (remitamine) [inline youtube_dl/extractor/samplefocus.py:38]: i did request multiple times to check the fallbacks, check that the code is working before pushing.

**Pull request closed without merging**